### PR TITLE
[Refactor] Decouple Orion

### PR DESCRIPTION
**Pull request by MDK5**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
Currently Orion specific buttons and logic is spread all over Artemis. Orion should be decoupled and have its own components and services.

Also happens to fix https://github.com/ls1intum/Orion/issues/54. It was caused by a duplicated \<jhi-alert\> that got removed by the duplication removal.

### Description
This PR introduces a new extensionPoint directive to allow Orion to reuse Artemis' templates and then decouples various components. In the future the Orion components also need to be moved to their own modules and get their own routing, this will be done at a later date; this PR only aims to divide the components.

There are no functional changes, so the tests only need to ensure there is no regression. Additionally the changes have been almost exclusively visual, so the focus should be ensuring every button is still show correctly and connected to the right function.

### Steps for Testing
The following pages should be visited with an Orion client and with a normal client and then be tested for a functioning GUI.
1. The exercise details page, especially the student actions (start/resume/clone exercise)
2. The instructor course exercise overview, especially the exercise counter at the top and the programming exercise table (Orion does not show the templates or build plans and overrides the action buttons in the last row)
3. [Orion only] the Orion instructor code editor (that opens when clicking the open in Orion in the overview in point 2)
4. The assessment exercise dashboard, especially for programming exercises.

For further reference take a look at the components in app/orion, all of those got changed in some way and should be tested

**Note for code reviews**: Most code just got moved, therefore the diff looks a bit ugly. If some method shows as removed it most likely will appear somewhere else, if some method just appears, it will probably be removed somewhere else, only exception being some added tests and some removed duplication from the student-exercise-actions.

### Test Coverage
Everything in /orion: 100%

#### Exercise details
Without Orion:
grafik
With Orion:
grafik

#### Exercise overview
Without Orion
grafik
With Orion:
grafik

#### Code Editor
Orion only
grafik

#### Assessment exercise dashboard
Without Orion
grafik
With Orion
grafik

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts
Comment (krusche) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:137]: add documentation

**Review by PatZintel — APPROVED** [review]
worked in a testing session on ts1

**Review by JohannesStoehr — APPROVED** [review]
Tested during testing session. All functions are still working in the browser!

File: src/main/webapp/app/course/manage/course-management-exercises.component.ts
Comment (FlorianGlombik) [inline src/main/webapp/app/course/manage/course-management-exercises.component.ts:36]:
```suggestion
     * @param count to set the programmingExerciseCount to
```

**Review by FlorianGlombik — DISMISSED** [review]
Tested as an instructor and the buttons were displayed and worked with Orion as well as with the browser.

**Review by jhainzlschmid — DISMISSED** [review]
Tested in the testing session. Looks good

**Review by JohannesStoehr — APPROVED** [review]
Re-approve after merge

**Review by FlorianGlombik — APPROVED** [review]
Re-approving after merge. 
I had another look at the assessment dashboard in Orion and on Artemis as an instructor and didn't notice unexpected behavior.

**Review by jhainzlschmid — DISMISSED** [review]
I also re-checked the assessment dashboard in Orion and everything looked good to me

File: src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts
Comment (Olimbia23) [reply src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts:508]: On second thought, I am not sure this method even belongs here. I think this method should be inside Submission class. Something like:
```js
export const enum SubmissionStatus {
  DRAFT = 'DRAFT',
  DONE = 'DONE'
}

export abstract class Submission implements BaseEntity {
  ...
  submissionStatusForCorrectionRound(correctionRound = 0): SubmissionStatus {
    ...
    return SubmissionStatus.DRAFT;
  }
}
```

File: src/main/webapp/app/orion/management/orion-course-management-exercises.component.ts
Comment (Olimbia23) [inline src/main/webapp/app/orion/management/orion-course-management-exercises.component.ts:5]: [nit] What do you think, is it better inline html here as the component is really small and has no typescript code? Like here: https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/shared/alert/alert.component.ts#L6-L17

File: src/main/webapp/app/orion/management/orion-programming-exercise.component.ts
Comment (Olimbia23) [inline src/main/webapp/app/orion/management/orion-programming-exercise.component.ts:38]: What is the reason this navigation can fail? Could you maybe add a comment for it?

File: src/main/webapp/app/orion/participation/orion-course-exercise-details.component.ts
Comment (Olimbia23) [inline src/main/webapp/app/orion/participation/orion-course-exercise-details.component.ts:9]: [nit] What do you think of inlining the html, as it is small and the component itself has no typescript code

**Review by Olimbia23 — APPROVED** [review]
Great work @MDK5. Left some suggestions 👍

File: src/test/javascript/spec/component/orion/orion-programming-exercise.component.spec.ts
Comment (FuchsDominik) [inline src/test/javascript/spec/component/orion/orion-programming-exercise.component.spec.ts:44]:
```suggestion
    afterEach(() => {
        router.navigateSpy.restore();
        sinon.restore();
    });
```

**Review by FuchsDominik — APPROVED** [review]
Code looks really good in my opinion, did not find any big issues. I would approve @Olimbia23's suggestions of moving the really small html template into the typescript component. Other than that we could call `sinon.restore()` after the test cases which should clear all stubs and **also the spys**.

File: src/main/webapp/app/orion/management/orion-programming-exercise.component.ts
Comment (MDK5) [reply src/main/webapp/app/orion/management/orion-programming-exercise.component.ts:38]: I have to admit I have no idea an have never seen any failures. I don't really feel comfortable with deleting it either though, assuming whoever put it there probably had their reasons.

File: src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts
Comment (MDK5) [reply src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts:508]: I now settled with a helper returning a boolean since the enum was not really needed in the first place. Good idea!

File: src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.html
Comment (FrankeLukas) [inline src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.html:68]: Is the white space between the tags intentional?

File: src/main/webapp/app/orion/assessment/orion-exercise-assessment-dashboard.component.html
Comment (FrankeLukas) [inline src/main/webapp/app/orion/assessment/orion-exercise-assessment-dashboard.component.html:2]: Same here.

File: src/main/webapp/app/orion/participation/orion-exercise-details-student-actions.component.html
Comment (FrankeLukas) [inline src/main/webapp/app/orion/participation/orion-exercise-details-student-actions.component.html:14]:
```suggestion
            [buttonLoading]="this.exercise.loading || orionState.cloning"
```

**Review by FrankeLukas — DISMISSED** [review]
Code changes lgtm otherwise. I'd also make `calculateSubmissionStatus` a method returning a boolean, but seeing your last comment you're already doing that?

File: src/main/webapp/app/orion/management/orion-course-management-exercises.component.ts
Comment (MDK5) [reply src/main/webapp/app/orion/management/orion-course-management-exercises.component.ts]: Makes sense, inlined the html :+1:

File: src/main/webapp/app/orion/participation/orion-course-exercise-details.component.ts
Comment (MDK5) [reply src/main/webapp/app/orion/participation/orion-course-exercise-details.component.ts]: Makes sense, inlined the html :+1:

File: src/test/javascript/spec/component/orion/orion-exercise-assessment-dashboard.component.spec.ts
Comment (MDK5) [reply src/test/javascript/spec/component/orion/orion-exercise-assessment-dashboard.component.spec.ts]: Thats a nice feature, did not know that yet. Thanks!

File: src/main/webapp/app/orion/management/orion-programming-exercise.component.ts
Comment (FuchsDominik) [reply src/main/webapp/app/orion/management/orion-programming-exercise.component.ts:38]: I would rather haven an additional check than an error which does not get caught :D

**Review by FrankeLukas — APPROVED** [review]
Comments were addressed. Haven't tested the functionality though.

**Review by FuchsDominik — APPROVED** [review]
Code-wise approve since change requests have been addressed

**Review by JohannesStoehr — APPROVED** [review]
The latest changes didn't break anything in Artemis.
Good work

**Review by jhainzlschmid — APPROVED** [review]
I retested the normal Artemis client and didn't find any issues.

**Review by FlorianGlombik — APPROVED** [review]
Tested the Artemis side as instructor:
- start + resume exercise worked
- all buttons in the exercise overview worked _(did not test resetting the exercise since I am not familiar with this functionality)_
- didn't notice unexpected behavior don't on the assessment dashboard page

**Pull request merged**
